Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,106 @@
+/-
+Copyright (c) 2024 Jz Pan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jz Pan
+-/
+import Mathlib.AlgebraicGeometry.EllipticCurve.Weierstrass
+
+/-!
+
+# Some normal forms of elliptic curves
+
+This file defines some normal forms of Weierstrass equations of elliptic curves.
+
+## Main definitions
+
+- ...
+
+## References
+
+* [J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, weierstrass equation, normal form
+
+-/
+
+universe u
+
+variable {R : Type u}
+
+namespace WeierstrassCurve
+
+variable (W : WeierstrassCurve R)
+
+/-! ### Normal forms of characteristic ≠ 2
+-/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic ≠ 2, if its `a₁` and `a₃` are zero.
+In other words it is $Y^2 = X^3 + a_2X^2 + a_4X + a_6$. -/
+@[mk_iff]
+structure IsNormalFormOfCharNeTwo [Zero R] : Prop where
+  a₁ : W.a₁ = 0
+  a₃ : W.a₃ = 0
+
+section VariableChange
+
+variable [CommRing R] [Invertible (2 : R)]
+
+/-- This is an explicit change of variables of a `WeierstrassCurve` to
+a normal form of characteristic ≠ 2, provided that 2 is invertible in the ring. -/
+@[simps]
+def variableChangeToNormalFormOfCharNeTwo : VariableChange R where
+  u := 1
+  r := 0
+  s := ⅟2 * -W.a₁
+  t := ⅟2 * -W.a₃
+
+theorem variableChangeToNormalFormOfCharNeTwo_spec :
+    (W.variableChange (W.variableChangeToNormalFormOfCharNeTwo)).IsNormalFormOfCharNeTwo := by
+  constructor <;> simp [← mul_assoc]
+
+theorem exists_variableChange_isNormalFormOfCharNeTwo :
+    ∃ C : VariableChange R, (W.variableChange C).IsNormalFormOfCharNeTwo :=
+  ⟨_, W.variableChangeToNormalFormOfCharNeTwo_spec⟩
+
+end VariableChange
+
+namespace IsNormalFormOfCharNeTwo
+
+variable {W}
+variable [CommRing R] (self : W.IsNormalFormOfCharNeTwo)
+include self
```

Comment:
Hmm I've never needed to do `include` before. Does making `IsNormalFormOfCharNeTwo` a typeclass help?

Reply:
It is a "feature" introduced in [4.11.0](https://github.com/leanprover/lean4/releases/tag/v4.11.0). In fact I'm refrained from making it as a typeclass...